rust/headless-client/src/windows.rs

```diff
@@ -29,3 +226,79 @@ pub(crate) fn default_token_path() -> std::path::PathBuf {
 pub(crate) fn notify_service_controller() -> Result<()> {
     Ok(())
 }
+
+#[cfg(test)]
+mod test {
+    use super::*;
+    use firezone_bin_shared::TunDeviceManager;
+    use ip_network::Ipv4Network;
+    use socket_factory::DatagramOut;
+    use std::borrow::Cow;
+    use std::net::{Ipv4Addr, Ipv6Addr, SocketAddr, SocketAddrV4};
+    use std::time::Duration;
+
+    #[test]
+    fn best_route_ip4_does_not_panic_or_segfault() {
+        let _ = get_best_non_tunnel_route("8.8.8.8".parse().unwrap());
+    }
+
+    #[test]
+    fn best_route_ip6_does_not_panic_or_segfault() {
+        let _ = get_best_non_tunnel_route("2404:6800:4006:811::200e".parse().unwrap());
+    }
+
+    // Starts up a WinTUN device, adds a "full-route" (`0.0.0.0/0`) and checks if we can still send packets to IPs outside of our tunnel.
+    #[tokio::test]
+    #[ignore = "Needs admin & Internet"]
+    async fn no_packet_loops() {
```

Comment:
@ReactorScram This is how I imagined we can get some more automated test coverage for platform-specific IO stuff.

Reply:
Yeah now that we have the tunnel device split out from the connlib / portal stuff it looks like it will do what I wanted